src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs

```diff
@@ -103,264 +103,146 @@ public CSharpCodeParser(IEnumerable<DirectiveDescriptor> directives, ParserConte
+            if (Context == null)
             {
-                _keywordParsers.Add(keyword, handler);
-                if (topLevel)
-                {
-                    Keywords.Add(CSharpLanguageCharacteristics.GetKeyword(keyword));
-                }
+                throw new InvalidOperationException(Resources.Parser_Context_Not_Set);
             }
-        }
-
-        [Conditional("DEBUG")]
-        internal void Assert(CSharpKeyword expectedKeyword)
-        {
-            var result = CSharpTokenizer.GetTokenKeyword(CurrentToken);
-            Debug.Assert(CurrentToken.Kind == SyntaxKind.Keyword &&
-                result.HasValue &&
-                result.Value == expectedKeyword);
-        }
 
-        protected internal bool At(CSharpKeyword keyword)
-        {
-            var result = CSharpTokenizer.GetTokenKeyword(CurrentToken);
-            return At(SyntaxKind.Keyword) &&
-                result.HasValue &&
-                result.Value == keyword;
-        }
-
-        protected internal bool AcceptIf(CSharpKeyword keyword)
-        {
-            if (At(keyword))
+            if (EndOfFile)
             {
-                AcceptAndMoveNext();
-                return true;
+                // Nothing to parse.
+                return null;
             }
-            return false;
-        }
-
-        protected static Func<SyntaxToken, bool> IsSpacingToken(bool includeNewLines, bool includeComments)
-        {
-            return token => token.Kind == SyntaxKind.Whitespace ||
-                          (includeNewLines && token.Kind == SyntaxKind.NewLine) ||
-                          (includeComments && token.Kind == SyntaxKind.CSharpComment);
-        }
 
-        public override void ParseBlock()
-        {
-            using (PushSpanConfig(DefaultSpanConfig))
+            using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+            using (PushSpanContextConfig(DefaultSpanContextConfig))
             {
-                if (Context == null)
-                {
-                    throw new InvalidOperationException(Resources.Parser_Context_Not_Set);
-                }
-
-                Span.Start = CurrentLocation;
-
-                // Unless changed, the block is a statement block
-                using (Context.Builder.StartBlock(BlockKindInternal.Statement))
+                var builder = pooledResult.Builder;
+                try
                 {
                     NextToken();
 
+                    // Unless changed, the block is a statement block
                     AcceptWhile(IsSpacingToken(includeNewLines: true, includeComments: true));
+                    builder.Add(OutputTokensAsStatementLiteral());
 
-                    var current = CurrentToken;
+                    // We are usually called when the other parser sees a transition '@'. Look for it.
+                    SyntaxToken transitionToken = null;
                     if (At(SyntaxKind.StringLiteral) &&
```

Comment:
Curious, why string literal here?

Reply:
Because our Tokenizer is weird and tokenizes `@""` as one StringLiteral token.